Avoid possible `rm -rf /` when running `make check`

iamleot opened the pull request:
- Always avoid trailing slash in `rm -rf $(var)/'. When `$(var)' is not
  initialized this can ends up in `rm -rf /'.  The trailing slash
  is not needed and can be just avoided.
- Avoid GNU_MAKE usage.
- Conditionalize WITH_APPLESPELL logic in tests for targets that are used
  only when WITH_APPLESPELL is defined.

This fixes #201 and probably also fixes #185.

iamleot committed:
Avoid trailing slash in `rm -rf $(var)/'

When removing via `rm -rf' adding a trailing slash can be dangerous
when `$(var)' is undefined and ends up in completely removing `/'
as happened in #185 and #201.
They are not needed and it is better to avoid them.

This avoid `rm -rf /' when running `make check'.

Please note that if the current user has permission to write a possible
`/enchant-@ENCHANT_MAJOR_VERSION@' directory that directory will be created
(i.e. the problem described by issue #201 is not completely addressed but this
commit but should avoid possible nasty `rm -rf')

rrthomas commented:
Thanks for these.

Please can you remove the second and third commits? The third one should go in a separate pull request; meanwhile, the second one is a bug in gnulib, which I'll report.

iamleot commented:
Reuben Thomas writes:
> Thanks for these.
>
> Please can you remove the second and third commits? The third one should go in a separate pull request; meanwhile, the second one is a bug in gnulib, which I'll report.

If the 2nd and 3rd commits will be removed it will ends up doing
a `rm -rf' so that's still broken.

Can you please provide a rationale regarding using GNU_MAKE and
the `$(shell ...)'? (skimming through `git blame' and `git log'
didn't helped)


Thank you!

rrthomas commented:
`rm -rf` exits with status 0, and does nothing, so that will be OK for starters.

GNU Make is used for `$(shell)` to compute `libdir_subdir`. See commit e95b1fdb.

iamleot commented:
Hello Reuben,

Reuben Thomas writes:
> `rm -rf` exits with status 0, and does nothing, so that will be OK for starters.
>
> GNU Make is used for `$(shell)` to compute `libdir_subdir`. See commit e95b1fdb.

The commit introducing that seems to be
a8e771a58a996d210dc9d12a3896d04b27465cc1.

Why libdir_subdir was introduced, couldn't the `$(shell ...)' be
avoided completely?

rrthomas commented:
`libdir_subdir` is needed to run the tests in-place.

iamleot commented:
Reuben Thomas writes:
> `libdir_subdir` is needed to run the tests in-place.

Couldn't the .la/.a/.so copied unconditionally in a lib/ directory
to avoid that? (so that all the g_module_open() calls and other
users of LIBDIR_SUBDIR will DTRT)

rrthomas commented:
Happy to have a patch if you think you can simplify things. Note: the files are already copied, but they still need to be found in the first place.

iamleot commented:
Reuben Thomas writes:
> Happy to have a patch if you think you can simplify things.

If I can, yes, I would be happy to too...

During the `make check' the following two commands fails (actually
it doesn't block the `check' target execution but probably it is
not expected):

 cp ../../src/libenchant-2.la lib/
 cp: ../../providers/.libs/*.so: No such file or directory
 PASS: providers.test
 [...]
 cp -r ../src/.libs lib/
 cp ../src/libenchant-2.la lib/
 cp: ./.libs/*.so: No such file or directory
 FAIL: main.test
 [...]

(The two `cp .libs/*.so' probably expect shared objects but couldn't
find any.)

I am not sure if this is related to the following warnings:

 *** Warning: linker path does not have real file for library -lgcc.
 *** I have the capability to make that library automatically link in when
 *** you link to this library.  But I can only do this if you have a
 *** shared version of the library, which you do not appear to have
 *** because I did check the linker path looking for a file starting
 *** with libgcc and none of the candidates passed a file format test
 *** using a regex pattern. Last file checked: /usr/lib/libgcc.a
 
 *** Warning: libtool could not satisfy all declared inter-library
 *** dependencies of module libenchant_mock_provider.  Therefore, libtool will create
 *** a static module, that should work as long as the dlopening
 *** application is linked with the -dlopen flag.

This happens for the following modules:

 - libenchant_mock_provider
 - libenchant_mock_provider2
 - libenchant_null_provider
 - libenchant_null_identify
 - libenchant_null_describe

And than when ktruss(1) `make check' I see that:

 - providers.test just opens libenchant-2.so.4 that is in
   tests/enchant_providers/lib/.libs/libenchant-2.so.4
 - main.test try to opens (but get ENOENT because none of them exists):
    - lib/enchant-2 (via various open() and rmdir())
    - lib/enchant-2/enchant_mock_provider{,.so,.la}

Is this expected?  How is this related to libdir_subdir that needs
to be extracted from libdir?  Can you please share the output of
`strace' or similar regarding the files that are opened by
`g_module_open()' and `AddToPath()' of the tests?
(they should be:
  - LIBDIR_SUBDIR "/enchant-" ENCHANT_MAJOR_VERSION "/enchant_mock_provider"
  - LIBDIR_SUBDIR "/enchant-" ENCHANT_MAJOR_VERSION "/enchant_mock_provider2"
 ...and probably other that should match `LIBDIR_SUBDIR "/enchant-"'
 concatenation though)


Thank you!

rrthomas commented:
I don't think exchanging low-level traces is the way to go about this (of course it's useful for debugging). But I think the question of what should happen is best addressed from the top down, not for example by my sharing strace output.

In the case of the missing `.so` files under `providers`, I guess that's what you would expect if you built without any providers. (I have one `.so` file per provider I build.)

The warnings you mention do look problematic: it sounds as though the required `.so` files will thereby not be created.

As for the expected locations of files, you can see that in `tests/Makefile.am`. The setting of `libdir_subdir` is a best effort attempt to recreate the system paths under the local directory.

gnulib has now been fixed, so the problem with NetBSD Make should no longer exist. Again, if you could change this pull request to include only the first commit, then we should have fixed two of the three problems.

iamleot commented:
Hello Reuben,

Reuben Thomas writes:
> [...]
> As for the expected locations of files, you can see that in `tests/Makefile.am`. The setting of `libdir_subdir` is a best effort attempt to recreate the system paths under the local directory.
>

Sorry but I still do not understand that. Why it is attempted to
recreate the system paths if the tests are in-place? (or, why it
is important that the libdir_subdir is derived from libdir if the
final `.so' should be loaded from tests/?)

> gnulib has now been fixed, so the problem with NetBSD Make should no longer exist.

Thank you for that!

> Again, if you could change this pull request to include only the first commit, then we should have fixed two of the three problems.

Yes, I've now adjusted the PR to only include the first commit and
adjusted the commit message accordingly.

rrthomas commented:
Sorry, I think I was unclear. The tests are made on an "in-place" install, which must have the same structure as the installed files (so that the relative paths the code constructs match).

rrthomas commented:
Thanks. Please can you open another PR with the AppleSpell logic in, and I can then try to understand that?

iamleot commented:
Thank you @rrthomas!

A possible analysis of this problem was shared in the last comment in #201.

Please let me know if more information/testing/feedback is needed.

rrthomas commented:
I don't understand the analysis. Why would `$<` be empty? It's the first prerequisite, which is always `$(top_srcdir)/providers/AppleSpell.config`.

iamleot commented:
Reuben Thomas writes:
> I don't understand the analysis. Why would `$<` be empty? It's the first prerequisite, which is always `$(top_srcdir)/providers/AppleSpell.config`.

I don't know, sorry. Not sure if it helps but invoking via
`make -d s check':

    [...]
    Wildcard expanding "Provider/provider_list_dicts.cpp"...suffix is ".cpp"...SuffFindDeps (Provider/provider_list_dicts.cpp)
    Wildcard expanding "Provider/provider_request_dict.cpp"...suffix is ".cpp"...SuffFindDeps (Provider/provider_request_dict.cpp)
    Wildcard expanding "../../src/libenchant-2.la"...SuffFindDeps (../../src/libenchant-2.la)
            No known suffix on ../../src/libenchant-2.la. Using .NULL suffix
    adding suffix rules
    Wildcard expanding "../../providers/AppleSpell.config"...SuffFindDeps (../../providers/AppleSpell.config)
            No known suffix on ../../providers/AppleSpell.config. Using .NULL suffix
    adding suffix rules
    Wildcard expanding "config/AppleSpell.config"...../../build-aux/install-sh -c -d config
    cp  config/
    [...]

rrthomas commented:
Thanks, but I'm not sure this helps me much. Does "Using .NULL suffix" mean it drops `../../providers/AppleSpell.config` altogether?

Note that this file always exists!

iamleot commented:
Reuben Thomas writes:
> Thanks, but I'm not sure this helps me much. Does "Using .NULL suffix" mean it drops `../../providers/AppleSpell.config` altogether?
>
> Note that this file always exists!

I think that it is not treated as a suffix-transformation rule and
`$<' (`.IMPSRC') is not defined in the `$(APPLESPELL_CONFIG)'
target.

The following Makefile snippet probably reproduces what happens
with NetBSD make(1) (`make') and GNU Make (`gmake'):

   % touch ../AppleSpell.config
   % cat Makefile
   APPLESPELL_CONFIG = config/AppleSpell.config
   $(APPLESPELL_CONFIG): ../AppleSpell.config
           @echo ".TARGET \$$@" $@
           @echo ".PREFIX \$$*" $*
           @echo ".IMPSRC \$$<" $<
           @echo ".MEMBER \$$%" $%
   % make
   .TARGET $@ config/AppleSpell.config
   .PREFIX $* config/AppleSpell.config
   .IMPSRC $<
   .MEMBER $%
   % gmake
   .TARGET $@ config/AppleSpell.config
   .PREFIX $*
   .IMPSRC $< ../AppleSpell.config
   .MEMBER $%

rrthomas commented:
Thanks very much for digging into this.

I didn't think it was a suffix-transformation rule (that's why I was puzzled by the messages about this).

I see what happens now, but I still don't understand why. This is a normal make rule, it defines a prerequisite; why is the normal prerequisite macro `$<` not defined?

iamleot commented:
Reuben Thomas writes:
> I see what happens now, but I still don't understand why. This is a normal make rule, it defines a prerequisite; why is the normal prerequisite macro `$<` not defined?

I think because the semantic of `$<' in GNU make is different.
In NetBSD make(1) `${.IMPSRC}' (`$<') is defined only in
suffix-transformation rule (I think in the GNU make info these are
`Old-Fashioned Suffix Rules').

This is not a suffix rule (and I think that it's the same case of
APPLESCRIPT_CONFIG rule):

    % cat Makefile
    foo: bar
            @echo $<
    % touch bar
    % make
    
    % gmake
    bar

While this is:

    % cat Makefile
    .SUFFIXES: .foo .baz
    
    .foo.baz: bar
            @echo $<
    % touch x.foo
    % make x.baz
    x.foo
    % gmake x.baz
    x.foo

And I guess that only in this case using `$<' is probably portable.

(Unfortunately I'll probably not be able to shed more light, sorry,
but feel free to ping me if you need any possible testing.)

rrthomas commented:
Ah, now I understand! Thanks, I'll look into this and prepare a fix.

rrthomas commented:
It seems that indeed `$<` is only defined by POSIX in inference rules, so my use of it is non-standard. I shall make a fix.
